aioxmpp/muc/self_ping.py

```diff
@@ -0,0 +1,429 @@
+        Thus, if the :attr:`ping_interval` was set to one day and is then
+        changed to one minute, it takes up to a day until the one minute
+        interval starts being used.
+        """
+        return self._ping_interval
+
+    @ping_interval.setter
+    def ping_interval(self, value: timedelta):
+        # cheap & duck-typey enforcement of timedelta compatibility
+        self._ping_interval = value + timedelta()
+
+    @property
+    def ping_timeout(self) -> timedelta:
+        """
+        The maximum time to wait for a reply to a ping.
+
+        Each ping sent by the pinger has its individual timeout, based on this
+        property at the time the ping is sent.
+        """
+        return self._ping_timeout
+
+    @ping_timeout.setter
+    def ping_timeout(self, value: timedelta):
+        # cheap & duck-typey enforcement of timedelta compatibility
+        self._ping_timeout = value + timedelta()
+
+    def start(self):
+        """
+        Start the pinging coroutine using the client and event loop which was
+        passed to the constructor.
+
+        :meth:`start` always behaves as if :meth:`stop` was called right before
+        it.
+        """
+        self.stop()
+        self._task = asyncio.ensure_future(self._pinger(), loop=self._loop)
+
+    def stop(self):
+        """
+        Stop the pinger (if it is running) and discard all data on in-flight
+        pings.
+
+        This method will do nothing if the pinger is already stopped. It is
+        idempotent.
+        """
+        if self._task is None:
+            return
+
+        self._task.cancel()
+        self._task = None
+
+    def _interpret_result(self, task):
+        """
+        Interpret the result of a ping.
+
+        :param task: The pinger task.
+
+        The result or exception of the `task` is interpreted as follows:
+
+        * :data:`None` result: *positive*
+        * :class:`aioxmpp.errors.XMPPError`, ``service-unavailable``: *positive*
+        * :class:`aioxmpp.errors.XMPPError`, ``feature-not-implemented``:
+          *positive*
+        * :class:`aioxmpp.errors.XMPPError`, ``item-not-found``: *inconclusive*
+        * :class:`aioxmpp.errors.XMPPError`: *negative*
+        * :class:`asyncio.TimeoutError`: *inconclusive*
+        * Any other exception: *inconclusive*
+        """
+        if task.exception() is None:
+            self._on_fresh()
+            return
+
+        exc = task.exception()
+        if isinstance(exc, aioxmpp.errors.XMPPError):
+            if exc.condition in [
+                    aioxmpp.errors.ErrorCondition.SERVICE_UNAVAILABLE,
+                    aioxmpp.errors.ErrorCondition.FEATURE_NOT_IMPLEMENTED]:
+                self._on_fresh()
+                return
+
+            if exc.condition == aioxmpp.errors.ErrorCondition.ITEM_NOT_FOUND:
+                return
+
+            self._on_exited()
+
+    @asyncio.coroutine
+    def _pinger(self):
```

Comment:
The resend timeout should be randomized (uniformly, probably in a range `[0.5..1.5]*base_timeout`) to avoid clustering of pings after a MUC got silent.

Reply:
This will screw up the tests baaaad. I have no idea how to do this while still having timing tests which run at least mostly reliable.

Although I see that this is a desired feature. Probably 10% jitter is enough though.

Reply:
Yes it probably is (the +- 50% jitter was inspired by RTCP).